homeassistant/components/light/hue.py

```diff
@@ -48,260 +41,215 @@
-[Hue component](https://home-assistant.io/components/hue/) page for more
-information.
-"""
-
-SIGNAL_CALLBACK = 'hue_light_callback_{}_{}'
 
-
-def setup_platform(hass, config, add_devices, discovery_info=None):
+async def async_setup_platform(hass, config, async_add_devices,
+                               discovery_info=None):
     """Set up the Hue lights."""
-    if discovery_info is None or 'bridge_id' not in discovery_info:
-        return
-
-    if config is not None and config:
-        # Legacy configuration, will be removed in 0.60
-        config_str = yaml.dump([config])
-        # Indent so it renders in a fixed-width font
-        config_str = re.sub('(?m)^', '      ', config_str)
-        hass.components.persistent_notification.async_create(
-            MIGRATION_INSTRUCTIONS.format(config=config_str),
-            title=MIGRATION_TITLE,
-            notification_id=MIGRATION_ID)
-
-    bridge_id = discovery_info['bridge_id']
-    bridge = hass.data[hue.DOMAIN][bridge_id]
-    unthrottled_update_lights(hass, bridge, add_devices)
-
-
-@util.Throttle(MIN_TIME_BETWEEN_SCANS, MIN_TIME_BETWEEN_FORCED_SCANS)
-def update_lights(hass, bridge, add_devices):
-    """Update the Hue light objects with latest info from the bridge."""
-    return unthrottled_update_lights(hass, bridge, add_devices)
-
-
-def unthrottled_update_lights(hass, bridge, add_devices):
-    """Update the lights (Internal version of update_lights)."""
-    import phue
-
-    if not bridge.configured:
+    if discovery_info is None:
         return
 
-    try:
-        api = bridge.get_api()
-    except phue.PhueRequestTimeout:
-        _LOGGER.warning("Timeout trying to reach the bridge")
-        bridge.available = False
-        return
-    except ConnectionRefusedError:
-        _LOGGER.error("The bridge refused the connection")
-        bridge.available = False
-        return
-    except socket.error:
-        # socket.error when we cannot reach Hue
-        _LOGGER.exception("Cannot reach the bridge")
-        bridge.available = False
-        return
-
-    bridge.available = True
-
-    new_lights = process_lights(
-        hass, api, bridge,
-        lambda **kw: update_lights(hass, bridge, add_devices, **kw))
-    if bridge.allow_hue_groups:
-        new_lightgroups = process_groups(
-            hass, api, bridge,
-            lambda **kw: update_lights(hass, bridge, add_devices, **kw))
-        new_lights.extend(new_lightgroups)
-
-    if new_lights:
-        add_devices(new_lights)
-
-
-def process_lights(hass, api, bridge, update_lights_cb):
-    """Set up HueLight objects for all lights."""
-    api_lights = api.get('lights')
-
-    if not isinstance(api_lights, dict):
-        _LOGGER.error("Got unexpected result from Hue API")
-        return []
+    bridge = hass.data[hue.DOMAIN][discovery_info['bridge_id']]
+    cur_lights = {}
+    cur_groups = {}
+
+    # Hue updates all lights via a single API call.
```

Comment:
I've had to make a choice between two update strategies. Neither are optimal. I've added an explanation as to what and why. I would love some feedback on that.

Reply:
Going to CC some people to get some extra insights: @pvizeli @MartinHjelmare @armills @amelchio @OttoWinter

Reply:
This sounds a lot like something reactive programming is trying to solve. So if I understand correctly, we're receiving lots of events with small delays and we somehow would like to "throttle" these events and bunch them together. 

One such filter is [`debounce`](https://github.com/ReactiveX/RxJava/wiki/Backpressure#debounce-or-throttlewithtimeout). It waits for the events to "settle down" a bit with some time delay and bunches the events. It does have its problems (like never timing out), but those can easily be fixed.

Reply:
I don't think it's safe to say that if an API call was already triggered but hasn't returned, a second update should just use the results from the first call. Even if it hasn't returned yet, the bridge might have already calculated what it's going to return.

I think the behavior we're looking for is that if an API call is currently in progress, lump together all the requests that are being made while the first call was in progress, and once the first call finishes, send a second call that will return for all the queued requests. This could be combined with a short debounce before any initial request to allow all the requests to be queued before sending the actual API call.

Reply:
> I don't think it's safe to say that if an API call was already triggered but hasn't returned, a second update should just use the results from the first call.

I think that in this case, it is safe to say that existing API calls will return the desired results. There are 2 scenarios when we call update on entities which both will be satisfied by `async_join_concurrent`:

 - We're polling all the entities and thus call update on each entity. All these update requests are all added to the event loop at once (because we do not specify a `PARALLEL_UPDATES` and the default for async platforms is all at once). In this scenario, joining on the existing request is fine because we all need the same update.

    The logic is equivalent to something like this:

    ```python
    await asyncio.wait(entity.async_update_ha_state(True)
                       for entity in entities if entity.should_poll)
    ```

 - We've just submitted a few commands to the hub and will call update on the entities afterwards. In this scenario too, just joining in on the first request is fine.

    The logic is equivalent to something like this:

    ```python
    for entity in entities:
        await entity.turn_on()
    
    await asyncio.wait(entity.async_update_ha_state(True)
                       for entity in entities if entity.should_poll)
    ```

The problem with debounce is that we need to specify a time to wait, which makes things slower.

Reply:
Alright, I fixed ! 

Now we are keeping track of which lights are waiting for the update to finish. Those lights will not have their update triggered.

Means that during a normal update cycle, none will be triggered. When we call a service on 2 entities, the other entities will get an update triggered, the other 2 will use their pending update.